src/OmniSharp.LanguageServerProtocol/Handlers/TextDocumentSyncHandler.cs

```diff
@@ -0,0 +1,196 @@
+        {
+            foreach (var (selector, openHandler, closeHandler, bufferHandler) in handlers
+                .OfType<
+                    Mef.IRequestHandler<FileOpenRequest, FileOpenResponse>,
+                    Mef.IRequestHandler<FileCloseRequest, FileCloseResponse>,
+                    Mef.IRequestHandler<UpdateBufferRequest, object>>())
+            {
+                // TODO: Fix once cake has working support for incremental
+                var documentSyncKind = TextDocumentSyncKind.Incremental;
+                // if (selector.ToString().IndexOf(".cake") > -1) documentSyncKind = TextDocumentSyncKind.Full;
+                yield return new TextDocumentSyncHandler(openHandler, closeHandler, bufferHandler, selector, documentSyncKind, workspace);
+            }
+        }
+
+        // TODO Make this configurable?
+        private readonly DocumentSelector _documentSelector;
+        private SynchronizationCapability _capability;
+        private readonly Mef.IRequestHandler<FileOpenRequest, FileOpenResponse> _openHandler;
+        private readonly Mef.IRequestHandler<FileCloseRequest, FileCloseResponse> _closeHandler;
+        private readonly Mef.IRequestHandler<UpdateBufferRequest, object> _bufferHandler;
+        private readonly OmniSharpWorkspace _workspace;
+
+        [ImportingConstructor]
+        public TextDocumentSyncHandler(
+            Mef.IRequestHandler<FileOpenRequest, FileOpenResponse> openHandler,
+            Mef.IRequestHandler<FileCloseRequest, FileCloseResponse> closeHandler,
+            Mef.IRequestHandler<UpdateBufferRequest, object> bufferHandler,
+            DocumentSelector documentSelector,
+            TextDocumentSyncKind documentSyncKind,
+            OmniSharpWorkspace workspace)
+        {
+            _openHandler = openHandler;
+            _closeHandler = closeHandler;
+            _bufferHandler = bufferHandler;
+            _workspace = workspace;
+            _documentSelector = documentSelector;
+            Options.Change = documentSyncKind;
+        }
+
+        public TextDocumentSyncOptions Options { get; } = new TextDocumentSyncOptions()
+        {
+            Change = TextDocumentSyncKind.Incremental,
+            OpenClose = true,
+            WillSave = false, // Do we need to configure this?
+            WillSaveWaitUntil = false,  // Do we need to configure this?
+            Save = new SaveOptions()
+            {
+                IncludeText = true
+            }
+        };
+
+        public TextDocumentAttributes GetTextDocumentAttributes(Uri uri)
+        {
+            var document = _workspace.GetDocument(Helpers.FromUri(uri));
+            if (document == null) return new TextDocumentAttributes(uri, "");
+            return new TextDocumentAttributes(uri, "");
+        }
+
+        public Task Handle(DidChangeTextDocumentParams notification)
+        {
+            if (notification.ContentChanges.Count() == 1 && notification.ContentChanges.First().Range == null)
```

Comment:
Why use an `IEnumerable` in the protocol when an array will do? When you use `IEnumerable`, things like this happen. :smile:

Reply:
Flexibility! 🏎 

(Converted to an array at the top of the method)

Reply:
How is an array (which implements `IEnumerable`) less flexible than `IEnumerable`? I would postulate that it's *more* flexible since it can be indexed.